### PR TITLE
Some more opinioted fixes

### PR DESCRIPTION
**Pull request by isidentical**
Please read the commits individually since there are some opinionated fixes (but I try to mainly follow PEP 701 text so these should be fine). Main goal was to actually fix all of `test_fstring` but I'm out of time so pushing this first batch. It brings down the number of failures from 28 to 12. (it also re-includes #61 but without any changes -- not sure what will be needed on that front since I can't reproduce failures).

### CODE REVIEW

**Comment by pablogsal** [comment]
CC: @lysnikolaou @mgmacias95

File: Lib/test/test_fstring.py [hunk 1]
```diff
@@ -825,7 +821,16 @@ def test_misformed_unicode_character_name(self):
                              r"'\N{GREEK CAPITAL LETTER DELTA'",
                              ])
 
-    def test_no_backslashes_in_expression_part(self):
+    def test_backslashes_in_expression_part(self):
```
Comment (MaxFirstQuadrant) [inline Lib/test/test_fstring.py:878]: maybe this should be renamed

**Comment by sunmy2019** [comment]
FWIW, Current failures of test_fstring


1. 
```
  File "<string>", line 1
    f'{"s"!r{":10"}}'
            ^
SyntaxError: invalid syntax

expected: "f-string: invalid conversion character 'r{"': expected 's', 'r', or 'a'"
```

2.
```
  File "<string>", line 1
    f'{4:{/5}}'
          ^
SyntaxError: invalid syntax

expected "f-string: invalid syntax" does not match "invalid syntax (<string>, line 1)"
```

3.
```
  File "?", line 1
    f'{a $ b}'
         ^
SyntaxError: invalid syntax

expected "f-string: invalid syntax"
````

4.
```
  File "<string>", line 1
    f'{lambda x:x}'
                ^
SyntaxError: invalid syntax

expected: "f-string: invalid syntax" 
```

5.
```
Traceback (most recent call last):
  File "/home/ubuntu/cpython-opinioted-fixes/Lib/test/test_fstring.py", line 32, in assertAllRaise
    with self.assertRaisesRegex(exception_type, regex):
AssertionError: SyntaxError not raised
```

6.
```
  File "<string>", line 1
    f'{3:'
         ^
SyntaxError: invalid syntax

expected: "f-string: expecting '}'"
```

7.
```
  File "<string>", line 1
    f'{3!s:'
           ^
SyntaxError: invalid syntax

expect: "f-string: expecting '}'"
```

8
```
  File "<string>", line 1
    f'{3!s:3'
            ^
SyntaxError: invalid syntax

expect: "f-string: expecting '}'"
```

9
```
  File "<string>", line 1
    f'{3:s'
          ^
SyntaxError: invalid syntax

expect: "f-string: expecting '}'" 
```

10
```
  File "<string>", line 1
    f'{!'
        ^
SyntaxError: f-string: expecting '}'

expect: "f-string: expression required before '!'" does not match "f-string: expecting '}' (<string>, line 1)"
```

11
```
  File "<string>", line 1
    f'{='
        ^
SyntaxError: f-string: expecting '}'

expect: "f-string: expression required before '='" does not match "f-string: expecting '}' (<string>, line 1)"
```

12
```
  File "?", line 1
SyntaxError: can't use starred expression here

expect: "cannot\ use\ starred\ expression\ here"
```

**Comment by sunmy2019** [comment]
Another notable failure is in `test_unparse`

The `ast.test_unparse` seems to output a merged `ast.JoinedStr`, causing the unparsed file have different ast.

Original
```py
        self.assertEqual('abc' f'## {x}ghi', 'abc## defghi')
        self.assertEqual('abc' f'{x}' 'ghi', 'abcdefghi')
        self.assertEqual('abc' f'{x}' 'gh' f'i{x:4}', 'abcdefghidef ')
        self.assertEqual('{x}' f'{x}', '{x}def')
        self.assertEqual('{x' f'{x}', '{xdef')
        self.assertEqual('{x}' f'{x}', '{x}def')
        self.assertEqual('{{x}}' f'{x}', '{{x}}def')
        self.assertEqual('{{x' f'{x}', '{{xdef')
        self.assertEqual('x}}' f'{x}', 'x}}def')
        self.assertEqual(f'{x}' 'x}}', 'defx}}')
        self.assertEqual(f'{x}' '', 'def')
        self.assertEqual('' f'{x}' '', 'def')
        self.assertEqual('' f'{x}', 'def')
        self.assertEqual(f'{x}' '2', 'def2')
        self.assertEqual('1' f'{x}' '2', '1def2')
        self.assertEqual('1' f'{x}', '1def')
        self.assertEqual(f'{x}' f'-{x}', 'def-def')
        self.assertEqual('' f'', '')
        self.assertEqual('' f'' '', '')
        self.assertEqual('' f'' '' f'', '')
        self.assertEqual(f'', '')
        self.assertEqual(f'' '', '')
        self.assertEqual(f'' '' f'', '')
        self.assertEqual(f'' '' f'' '', '')
```
unparsed
```py
        self.assertEqual(f'abc## {x}ghi', 'abc## defghi')
        self.assertEqual(f'abc{x}ghi', 'abcdefghi')
        self.assertEqual(f'abc{x}ghi{x:4}', 'abcdefghidef ')
        self.assertEqual(f'{{x}}{x}', '{x}def')
        self.assertEqual(f'{{x{x}', '{xdef')
        self.assertEqual(f'{{x}}{x}', '{x}def')
        self.assertEqual(f'{{{{x}}}}{x}', '{{x}}def')
        self.assertEqual(f'{{{{x{x}', '{{xdef')
        self.assertEqual(f'x}}}}{x}', 'x}}def')
        self.assertEqual(f'{x}x}}}}', 'defx}}')
        self.assertEqual(f'{x}', 'def')
        self.assertEqual(f'{x}', 'def')
        self.assertEqual(f'{x}', 'def')
        self.assertEqual(f'{x}2', 'def2')
        self.assertEqual(f'1{x}2', '1def2')
        self.assertEqual(f'1{x}', '1def')
        self.assertEqual(f'{x}-{x}', 'def-def')
        self.assertEqual(f'', '')
        self.assertEqual(f'', '')
        self.assertEqual(f'', '')
        self.assertEqual(f'', '')
        self.assertEqual(f'', '')
        self.assertEqual(f'', '')
        self.assertEqual(f'', '')
```

This is actually a huge behavior change.

**Comment by pablogsal** [comment]
Unfortunately I will be unavailable as I will be in Helsinki this weekend for heavy metal reasons,

@lysnikolaou can you review this one?

I gave you both permissions over my fork :)

**Comment by sunmy2019** [comment]
> Another notable failure is in `test_unparse`
> 
> ~The `ast.test_unparse` seems to output a merged `ast.JoinedStr`, causing the unparsed file have different ast.~
> 
> After investigation, the following test fails
> 
> ```python
> ast.dump(ast.parse("'' f'{x}'"))
> ```
> 
> expect
> 
> ```python
> "Module(body=[Expr(value=JoinedStr(values=[FormattedValue(value=Name(id='x', ctx=Load()), conversion=-1)]))], type_ignores=[])"
> ```
> 
> got
> 
> ```python
> "Module(body=[Expr(value=JoinedStr(values=[Constant(value=''), FormattedValue(value=Name(id='x', ctx=Load()), conversion=-1)]))], type_ignores=[])"
> ```
> 
> compile time folding is not working

branch fstring-grammar-rebased-after-sprint works fine.

File: Grammar/python.gram [hunk 2]
```diff
@@ -1362,3 +1362,7 @@ invalid_replacement_field:
     | '{' a=':' { RAISE_SYNTAX_ERROR_KNOWN_LOCATION(a, "f-string: expression required before ':'") }
     | '{' a='!' { RAISE_SYNTAX_ERROR_KNOWN_LOCATION(a, "f-string: expression required before '!'") }
     | '{' a='}' { RAISE_SYNTAX_ERROR_KNOWN_LOCATION(a, "f-string: empty expression not allowed") }
+    | '{' (yield_expr | star_expressions) "="? invalid_conversion_character
+invalid_conversion_character:
+    | a="!" &(':'|'}') { RAISE_SYNTAX_ERROR_KNOWN_LOCATION(a, "f-string: missed conversion character") }
```
Comment (lysnikolaou) [inline Grammar/python.gram:1367]: maybe this is slightly better?


```suggestion
    | a="!" &(':'|'}') { RAISE_SYNTAX_ERROR_KNOWN_LOCATION(a, "f-string: missing conversion character") }
```

**Review by lysnikolaou — APPROVED** [review]
Just one nit, looks great otherwise!

File: Grammar/python.gram (hunk 2 above)
Comment (pablogsal) [reply Grammar/python.gram:1367]: I will fix this in another push. Thanks for the suggestion!

**Pull request merged**